Update the Wallet FFI backend to persist the Comms Secret Key

philipr-za opened the pull request:
## Description
Previously the frontend client of the Wallet FFI library was responsible for persisting the Comms Secret Key. This made backing up and restoring the wallet tricky and as such it was decided that this private key should be persisted in the backend sqlite db.

This PR updates wallet backend DB with a new table called `wallet_settings` which will be a key-value store of wallet settings. The FFI function `comms_config_create(…)` to first check the wallet backend to see if a Comms Private Key is stored. If it is then it is used else a new one is generated.  The `wallet_create(…)` function will write whatever Comms Private Key is provided to it to the db for persistence.

In order to facilitate the migration to this new setup a new function is added to the FFI, `comms_config_set_secret_key(…)` which allows the frontend to specify the Comms Secret Key used. When migrating the Comms Config will be created. This will generate a new key because none will be found in the backend and then the frontend persisted Secret key will be set using the `comms_config_set_secret_key(…)` function and there after the backend will persist this secret key.

Tagging @kukabi and @Jasonvdm to note the required migration to happen in the frontends.

## Motivation and Context
Closes https://github.com/tari-project/tari/issues/1979

## How Has This Been Tested?
Test was added to exercise the updates to the `comms_config_create`, `comms_config_set_secret_key` and 'wallet_create' functions.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [ ] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [ ] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

kukabi reviewed (dismissed):
Tested it with an FFI build on this branch, working good.

sdbondi commented on base_layer/wallet_ffi/src/lib.rs:
```diff
@@ -2225,14 +2230,14 @@ pub unsafe extern "C" fn transport_type_destroy(transport: *mut TariTransportTyp
 /// `public_address` - The public address char array pointer. This is the address that the wallet advertises publicly to
 /// peers
-/// `listener_address` - The listener address char array pointer. This is the address that inbound peer
+/// `transport_type` - TariTransportType that specifies the expected format of the `public_address`
```
```suggestion
/// `transport_type` - TariTransportType that specifies the type of comms transport to be used.
```

sdbondi reviewed (commented):
LGTM - Just some 🖇️  to fix

StriderDM commented on base_layer/wallet_ffi/src/lib.rs:
```diff
@@ -2331,6 +2370,58 @@ pub unsafe extern "C" fn comms_config_create(
+/// # Safety
+/// None
+#[no_mangle]
+pub unsafe extern "C" fn comms_config_set_secret_key(
```
```
pub unsafe extern "C" fn comms_config_set_secret_key(
    comms_config: *const TariCommsConfig,
    secret_key: *const TariPrivateKey,
    error_out: *mut c_int,
)
```

philipr-za replied:
no the config needs to be mut because this function modifies it.

StriderDM replied:
Ooops! Meant `secret_key` should be `*const`

philipr-za replied:
done

StriderDM reviewed (dismissed):
LGTM

philipr-za committed:
Have the Wallet backend persist the Comms Secret Key

Previously the frontend client of the Wallet FFI library was responsible for persisting the Comms Secret Key. This made backing up and restoring the wallet tricky and as such it was decided that this private key should be persisted in the backend sqlite db.

This PR updates wallet backend DB with a new table called `wallet_settings` which will be a key-value store of wallet settings. The FFI function `comms_config_create(…)` to first check the wallet backend to see if a Comms Private Key is stored. If it is then it is used else a new one is generated.  The `wallet_create(…)` function will write whatever Comms Private Key is provided to it to the db for persistence.

In order to facilitate the migration to this new setup a new function is added to the FFI, `comms_config_set_secret_key(…)` which allows the frontend to specify the Comms Secret Key used. When migrating the Comms Config will be created. This will generate a new key because none will be found in the backend and then the frontend persisted Secret key will be set using the `comms_config_set_secret_key(…)` function and there after the backend will persist this secret key.
